### ISSUE TITLE
Unable to verify secret hash for client in v2.13.1 too but working fine for iOS

### ISSUE THREAD

**Issue by usmanrana07** [aws-amplify/aws-sdk-android#889]
I'm implementing **CustomAuthChallenge**'. This issue occurs when:

-  I try to pass user name for getting user session which is lets say 'abc_xyz'
-  Then it comes in '`getAuthenticationDetails`' method of the callback where i set username and auth type.
-  Then '`authenticateChallenge`' method is called in which 'continuation' object is received which has  a username in its params and that username is not the same that i passed but is userId '12121' of that user. 
- I set ANSWER in 'authenticateChallenge' and let it continueTask
- '`onFailure`' is called with the error 'Unable to verify secret hash for client'

 I checked the sdk and i saw  that userId is treated as '**usernameInternal**' . If i use userId '12121' in replacement of username 'abc_xyz' instead then it works fine. But in iOS it is working with username and we don't want to disclose userId so using username is mandatory. How can i resolve this issue?

**Comment by usmanrana07** [comment]
when is this expected to be resolved?

**Comment by charandas** [comment]
@usmanrana07 what's interesting is that `secretHash` can be calculated with `userId` for `getAuthenticationDetails`, it allows you to override (not that you need this capability):

```
           override fun getAuthenticationDetails(authContinuation: AuthenticationContinuation, username: String) {
                val authParameters = mutableMapOf<String, String>()
                authParameters[CognitoServiceConstants.AUTH_PARAM_USERNAME] = username
                val hash = CognitoSecretHash.getSecretHash(
                    authParams?.userId,
                    CognitoUserPoolHelper.CLIENT_ID,
                    CognitoUserPoolHelper.CLIENT_SECRET
                )

                val authenticationDetails = AuthenticationDetails(authParams?.userId, authParameters, null)
                authenticationDetails.authenticationType = CognitoServiceConstants.AUTH_TYPE_INIT_CUSTOM_AUTH
                authenticationDetails.customChallenge = CognitoServiceConstants.CHLG_TYPE_CUSTOM_CHALLENGE
                authenticationDetails.setAuthenticationParameter(CognitoServiceConstants.AUTH_PARAM_SECRET_HASH, hash)
                authenticationDetails.setAuthenticationParameter(CognitoServiceConstants.CHLG_RESP_SECRET_HASH, hash)
                authContinuation.setAuthenticationDetails(authenticationDetails)
                authContinuation.continueTask()
            }
```

The same cannot be said of `challengeContinuation` (now the overriding capability could get you past this issue 😉, but there is no such capability). With `challengeContinuation` it would read it out of the user. The user would set it up using `userId`. I have noticed that the backend does not want this.

```
    public CognitoUser getUser(String userId) {

        if (userId == null) {
            return getUser();
        }

        if (userId.isEmpty()) {
            return getUser();
        }

        return new CognitoUser(this, userId, clientId, clientSecret,
                CognitoSecretHash.getSecretHash(userId, clientId, clientSecret),
                client, context);
    }
```

The only silver lining is a user one sets up with `userPool.getUser()` instead of `getUser(userId)`. The latter sets secretHash to `null`, and so ends up benefiting from this `null` check in `updateInternalName`.

```
    private void updateInternalUsername(Map<String, String> challengeParameters) {
        if (usernameInternal == null) {
            if (challengeParameters != null && challengeParameters
                    .containsKey(CognitoServiceConstants.CHLG_PARAM_USERNAME)) {
                usernameInternal = challengeParameters
                        .get(CognitoServiceConstants.CHLG_PARAM_USERNAME);
                deviceKey = CognitoDeviceHelper.getDeviceKey(usernameInternal, pool.getUserPoolId(),
                        context);
                if (secretHash == null) {
                    secretHash = CognitoSecretHash.getSecretHash(usernameInternal, clientId,
                            clientSecret);
                }
            }
        }
    }
```

However, I run into other issues with this `getUser()`, such as not being able to call `updateUserAttributes` successfully after authenticating. Errors like `User-Id is null` come back.

[Here](https://github.com/aws-amplify/aws-sdk-android/issues/511#issuecomment-417147332) @kvasukib loosely recommends using a public (as opposed to confidential client) for custom auth. I am kinda concerned (security-wise) about this for this mobile app I am working out.

**Comment by charandas** [comment]
My settings are like so. I wonder if people are not using username attributes like phone number, if this ends up working for them. I am using the phone number as one.

<img width="1197" alt="Screen Shot 2019-05-15 at 3 11 17 PM" src="https://user-images.githubusercontent.com/542168/57802351-bbcf5a80-7723-11e9-8460-1e16833bc6a9.png">

**Comment by kvasukib** [comment]
We are adding this issue to our backlog. We will post an update here when we get to it.

**Comment by desokroshan** [comment]
A fix has been released in the version 2.15.2 of the SDK. Please feel free to upgrade and verify.
